### DISCUSSION TITLE
shared_ptr data interaction with spark or dask

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#190] [Design & Development]
I ran across a very subtle issue involving a standard (since C11) construct in C++, pybind11, and how spark or dask would handle a memory management issue.  This came up in trying to implement a "TopMute" class in C++ that utilizes our existing subclasses of BasicTaper.   Why is of secondary importance, but fyi the pwmig conversion requires a top mute operator and I was trying to "kill two birds with one stone" as a mutes are a stock seismic processing algorithm.

Ok, here is the subtle issue.  An implementation detail is that I was trying to use an inheritance mechanism that led me to write this little construct:
```
private:
  shared_ptr<BasicTaper> taper;
```
Why I used this construct is not important.  I can work around it if my suspicions prove true.  Understanding this issue is important, however, for C++ implementations in mspass so rather than do a workaround I thought I'd post this for discussion.

Normally when you use a shared pointer like that I could use this construct for the copy constructor (similar for operator=):
```
TopMute::TopMute(const TopMute& parent)
{
  this->taper=parent.taper;
}
```
The shared_ptr uses the now standard counting mechanism to free the memory of the original "taper" when the last pointer reference goes out of scope.   Python uses that same mechanism.

With that long introduction I can get to the subtle point.  If a job defined a TopMute that used that construct,  will it cause problems in spark and dask?   The thing the shared pointer defines is a block of memory but what happens when the image of the state is cloned to a different processor than the one that created the original instance of TopMute being passed around?   This is, I think, where the pickle definitions are needed on data.  This is different, however, because multiple instances may point at the same block of data and I don't have a clue how all that would be resolved?   More generally it is an issue of how implementations that use pointers like this handle that?   I suspect similar things would be problematic (or worse) with raw pointers.  I have not been able to find a readable source in this topic.   Do either of you know how this is handled?  If so, please educate me.

**Comment by wangyinz** [comment]
The key to your question is that both spark and dask should be treated as multiprocessing schedulers, so there is actually no data shared between tasks. That's also the reason why the map or reduce calls operates on functions or callable objects. It seems the `TopMute` class you showed here should be the latter.

**Reply by pavlis** [reply]
Thanks for the quick response.  My answer is slow because I've been out of internet contact for 2 days.  

Followup is probably in documentation, but call it a teachable moment:  In a python script that is using spark or dask, what symbols need to be be serializable?  Is it all data linked to a symbol defined in a script or only those that appear in parallel constructs (mainly map and reduce that is)?  

Alternatively, is it like MPI where when mpi is initiated the result looks like the old fashioned unix fork but with N branches (N being number of processes defined) instead of just 2?  If that is the model, I presume that fragmentation occurs occurs when a parallel operation is initiated.  Excuse my ignorance - this is a very elementary question I know but I realized I have some a fundamental lack of understanding here.

**Reply by pavlis** [reply]
PS to previous:   I just perused the DASK documentation and reminded why I'm confused on this question.  I need a good reference to teaching material on this topic instead of online references.  Wish this town still had a decent bookstore.  I may need to visit the campus bookstore on Monday when I go to the office.

**Reply by wangyinz** [reply]
Only those that appear in parallel constructs need to be serializable. If the object has some dependencies (such as importing certain modules), dask and spark will also take care of that. It is different from the MPI programs that you described, which is actually using a bulk synchronous parallel model. MapReduce is really like a master-worker approach, when a parallel operation is called, the master will send the tasks (as well as the data related to the tasks) to the workers.  So, the python script that a user wrote is only executed by the master, but the tasks generated from it are executed by the workers.

**Comment by pavlis** [comment]
Very helpful, thank you

On Apr 30, 2021, at 7:28 PM, Ian Wang ***@***.******@***.***>> wrote:

This message was sent from a non-IU address. Please exercise caution when clicking links or opening attachments from external sources.



Only those that appear in parallel constructs need to be serializable. If the object has some dependencies (such as importing certain modules), dask and spark will also take care of that. It is different from the MPI programs that you described, which is actually using a bulk synchronous parallel model. MapReduce is really like a master-worker approach, when a parallel operation is called, the master will send the tasks (as well as the data related to the tasks) to the workers. So, the python script that a user wrote is only executed by the master, but the tasks generated from it are executed by the workers.
